File: frontend/lib/features/recipe/models/recipe.dart
//mirrors the recipes table
//Lightweight summary shared with vault; ingredients/steps are null on the
//list endpoint and populated on the detail endpoint.
import 'package:mealchemy/features/recipe/models/recipe_ingredient.dart';
import 'package:mealchemy/features/recipe/models/recipe_step.dart';

class Recipe {
  final int recipeId;
  final int? ownerId;
  final String title;
  final String? description;
  final String? cuisineType;
  final int? prepTimeMins;
  final int? cookingTimeMins;
  final int? servingSize;
  final String? photoUrl;
  final String? videoUrl;
  final String? externalUrl;
  final bool isCommunityPublished;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  //populated by GET /recipes/{id}, null on list responses
  final List<RecipeIngredient>? ingredients;
  final List<RecipeStep>? steps;

  const Recipe({
    required this.recipeId,
    this.ownerId,
    required this.title,
    this.description,
    this.cuisineType,
    this.prepTimeMins,
    this.cookingTimeMins,
    this.servingSize,
    this.photoUrl,
    this.videoUrl,
    this.externalUrl,
    this.isCommunityPublished = false,
    this.createdAt,
    this.updatedAt,
    this.ingredients,
    this.steps,
  });

  factory Recipe.fromJson(Map<String, dynamic> json) {
    return Recipe(
      recipeId: json['recipeId'] as int,
      ownerId: json['ownerId'] as int?,
      title: json['title'] as String,
      description: json['description'] as String?,
      cuisineType: json['cuisineType'] as String?,
      prepTimeMins: json['prepTimeMins'] as int?,
      cookingTimeMins: json['cookingTimeMins'] as int?,
      servingSize: json['servingSize'] as int?,
      photoUrl: json['photoUrl'] as String?,
      videoUrl: json['videoUrl'] as String?,
      externalUrl: json['externalUrl'] as String?,
      isCommunityPublished: json['isCommunityPublished'] as bool? ?? false,
      createdAt: json['createdAt'] == null
          ? null
          : DateTime.parse(json['createdAt'] as String),
      updatedAt: json['updatedAt'] == null
          ? null
          : DateTime.parse(json['updatedAt'] as String),
      ingredients: (json['ingredients'] as List<dynamic>?)
          ?.map((e) => RecipeIngredient.fromJson(e as Map<String, dynamic>))
          .toList(),
      steps: (json['steps'] as List<dynamic>?)
          ?.map((e) => RecipeStep.fromJson(e as Map<String, dynamic>))
          .toList(),
    );
  }
  Map<String, dynamic> toFullRequestJson({bool removePhoto = false}) => {
        'title': title,
        'description': description ?? '',
        'cuisineType': cuisineType,
        'prepTimeMins': prepTimeMins,
        'cookingTimeMins': cookingTimeMins,
        'servingSize': servingSize,
        'photoUrl': photoUrl,
        'removePhoto': removePhoto,
        'videoUrl': videoUrl,
        'externalUrl': externalUrl,
        'isCommunityPublished': isCommunityPublished,
        'ingredients': ingredients?.map((i) => i.toJson()).toList() ?? [],
        'steps': steps?.map((s) => s.toJson()).toList() ?? [],
      };

//metadata-only body for POST /recipes/create and PUT /recipes/edit/{id}
  Map<String, dynamic> toCreateRequestJson() => {
        'title': title,
        'description': description,
        'cuisineType': cuisineType,
        'prepTimeMins': prepTimeMins,
        'cookingTimeMins': cookingTimeMins,
        'servingSize': servingSize,
        'photoUrl': photoUrl,
        'videoUrl': videoUrl,
        'externalUrl': externalUrl,
        'isCommunityPublished': isCommunityPublished,
      };
      
  Recipe copyWith({
    int? recipeId,
    int? ownerId,
    String? title,
    String? description,
    String? cuisineType,
    int? prepTimeMins,
    int? cookingTimeMins,
    int? servingSize,
    String? photoUrl,
    String? videoUrl,
    String? externalUrl,
    bool? isCommunityPublished,
    DateTime? createdAt,
    DateTime? updatedAt,
    List<RecipeIngredient>? ingredients,
    List<RecipeStep>? steps,
  }) {
    return Recipe(
      recipeId: recipeId ?? this.recipeId,
      ownerId: ownerId ?? this.ownerId,
      title: title ?? this.title,
      description: description ?? this.description,
      cuisineType: cuisineType ?? this.cuisineType,
      prepTimeMins: prepTimeMins ?? this.prepTimeMins,
      cookingTimeMins: cookingTimeMins ?? this.cookingTimeMins,
      servingSize: servingSize ?? this.servingSize,
      photoUrl: photoUrl ?? this.photoUrl,
      videoUrl: videoUrl ?? this.videoUrl,
      externalUrl: externalUrl ?? this.externalUrl,
      isCommunityPublished: isCommunityPublished ?? this.isCommunityPublished,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
      ingredients: ingredients ?? this.ingredients,
      steps: steps ?? this.steps,
    );
  }
}

File: backend/src/main/java/com/mealchemy/recipe/model/Recipe.java
package com.mealchemy.recipe.model;

/* Import libraries */

import jakarta.persistence.*;
import java.time.OffsetDateTime;
import java.util.*;
import org.hibernate.annotations.*;

/* Import Classes */

@Entity
@Table(name = "recipes")
public class Recipe
{
    /* Declaring fields */
    
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "recipe_id")
    private int recipeId;

    @Column(name = "owner_id", nullable = false)
    private Integer ownerId;

    @Column(name = "title", length = 200, nullable = false)
    private String title;

    @Column(name = "description", nullable = false)
    private String description;

    @Column(name = "cuisine_type", length = 40, nullable = false)
    private String cuisineType;

    @Column(name = "prep_time_mins", nullable = false)
    private int prepTimeMins;

    @Column(name = "cooking_time_mins", nullable = false)
    private int cookingTimeMins;

    @Column(name = "serving_size", nullable = false)
    private int servingSize;

    @Column(name = "photo_url", nullable = true)
    private String photoUrl;

    @Column(name = "video_url", nullable = true)
    private String videoUrl;

    @Column(name = "external_url", nullable = true)
    private String externalUrl;

    @Column(name = "is_community_published", nullable = false)
    private boolean isCommunityPublished = false;

    @CreationTimestamp
    @Column(name = "created_at", nullable = false)
    private OffsetDateTime createdAt = OffsetDateTime.now();

    @UpdateTimestamp
    @Column(name = "updated_at", nullable = false)
    private OffsetDateTime updatedAt = OffsetDateTime.now();

    @OneToMany(mappedBy = "recipe", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<RecipeIngredient> ingredients = new ArrayList<>();

    @OneToMany(mappedBy = "recipe", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<RecipeStep> steps = new ArrayList<>();

    @ManyToOne
    @JoinColumn(name = "parent_recipe_id")
    private Recipe parentRecipe;

    /* Getters */
    
    public int getRecipeId()
    {
        return recipeId;
    }

    public Integer getOwnerId()
    {
        return ownerId;
    }

    public String getTitle()
    {
        return title;
    }

    public String getDescription()
    {
        return description;
    }

    public String getCuisineType()
    {
        return cuisineType;
    }

    public int getPrepTimeMins()
    {
        return prepTimeMins;
    }

    public int getCookingTimeMins()
    {
        return cookingTimeMins;
    }

    public int getServingSize()
    {
        return servingSize;
    }

    public String getPhotoUrl()
    {
        return photoUrl;
    }

    public String getVideoUrl()
    {
        return videoUrl;
    }

    public String getExternalUrl()
    {
        return externalUrl;
    }

    public boolean getIsCommunityPublished()
    {
        return isCommunityPublished;
    }

    public OffsetDateTime getCreatedAt()
    {
        return createdAt;
    }

    public OffsetDateTime getUpdatedAt()
    {
        return updatedAt;
    }

    public List<RecipeIngredient> getIngredients()
    {
        return ingredients;
    }

    public List<RecipeStep> getSteps()
    {
        return steps;
    }

    public Recipe getParentRecipe()
    {
        return parentRecipe;
    }

    /* Setters */

    public void setOwnerId(Integer ownerIdIn)
    {
        ownerId = ownerIdIn;
    }

    public void setTitle(String titleIn)
    {
        title = titleIn;
    }

    public void setDescription(String descriptionIn)
    {
        description = descriptionIn;
    }

    public void setCuisineType(String cuisineTypeIn)
    {
        cuisineType = cuisineTypeIn;
    }

    public void setPrepTimeMins(int prepTimeMinsIn)
    {
        prepTimeMins = prepTimeMinsIn;
    }

    public void setCookingTimeMins(int cookingTimeMinsIn)
    {
        cookingTimeMins = cookingTimeMinsIn;
    }

    public void setServingSize(int servingSizeIn)
    {
        servingSize = servingSizeIn;
    }

    public void setPhotoUrl(String photoUrlIn)
    {
        photoUrl = photoUrlIn;
    }

    public void setVideoUrl(String videoUrlIn)
    {
        videoUrl = videoUrlIn;
    }

    public void setExternalUrl(String externalUrlIn)
    {
        externalUrl = externalUrlIn;
    }

    public void setIsCommunityPublished(boolean isCommunityPublishedIn)
    {
        isCommunityPublished = isCommunityPublishedIn;
    }

    public void setUpdatedAt(OffsetDateTime updatedAtIn)
    {
        updatedAt = updatedAtIn;
    }

    public void setIngredients(List<RecipeIngredient> ingredientsIn)
    {
        ingredients = ingredientsIn;
    }

    public void setSteps(List<RecipeStep> stepsIn)
    {
        steps = stepsIn;
    }

    public void setParentRecipe(Recipe parentRecipeIn)
    {
        parentRecipe = parentRecipeIn;
    }
}